File: app/src/main/java/com/rkbapps/neetflix/activityes/TvSeriesPreviewActivity.java
package com.rkbapps.neetflix.activityes;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager.widget.ViewPager;

import com.bumptech.glide.Glide;
import com.google.android.material.tabs.TabLayout;
import com.rkbapps.neetflix.R;
import com.rkbapps.neetflix.adapter.GenreAdapter;
import com.rkbapps.neetflix.adapter.TabLayoutAdapter;
import com.rkbapps.neetflix.models.tvseries.TvSeriesModel;
import com.rkbapps.neetflix.services.ApiData;
import com.rkbapps.neetflix.services.RetrofitInstance;
import com.rkbapps.neetflix.services.TvSeriesApi;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class TvSeriesPreviewActivity extends AppCompatActivity {

    private Toolbar toolbar;
    private TextView ratting, firstAirDate, originalLanguage, totalEpisodes, totalSeasons, tagLine;
    private RecyclerView recyclerGenre;
    private ViewPager viewPager;
    private TabLayout tabLayout;

    private ImageView backdrop;

    private TvSeriesModel tvSeriesModel;

    @SuppressLint("MissingInflatedId")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_tv_series_preview);
        int id = getIntent().getIntExtra("id", -1);
        String tittle = getIntent().getStringExtra("tittle");

        toolbar = findViewById(R.id.toolbarTvSeries);
        ratting = findViewById(R.id.txtRattingTvSeries);
        firstAirDate = findViewById(R.id.txtReleaseDateTvSeries);
        originalLanguage = findViewById(R.id.txtOriginalLanguageSeries);
        totalEpisodes = findViewById(R.id.txtNumberOfEpisodeSeries);
        totalSeasons = findViewById(R.id.txtTotalSeasons);
        tagLine = findViewById(R.id.txtTagLineTvSeries);
        recyclerGenre = findViewById(R.id.recyclerGenreTvSeries);
        viewPager = findViewById(R.id.pagerSeries);
        tabLayout = findViewById(R.id.tabLayoutSeries);
        backdrop = findViewById(R.id.imgBackdropTvSeries);

        recyclerGenre.setLayoutManager(new LinearLayoutManager(this, RecyclerView.HORIZONTAL, false));


        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle(tittle);


        TvSeriesApi api = RetrofitInstance.getTvSeriesApi();
        Call<TvSeriesModel> responseCall = api.getSeriesDetails(id, ApiData.API_KEY);
        responseCall.enqueue(new Callback<TvSeriesModel>() {
            @SuppressLint("SetTextI18n")
            @Override
            public void onResponse(Call<TvSeriesModel> call, Response<TvSeriesModel> response) {
                if (response.isSuccessful()) {
                    tvSeriesModel = response.body();

                    ratting.setText("" + tvSeriesModel.getVoteAverage());
                    firstAirDate.setText(tvSeriesModel.getFirstAirDate());
                    originalLanguage.setText(tvSeriesModel.getOriginalLanguage());
                    totalEpisodes.setText(tvSeriesModel.getNumberOfEpisodes() + "E");
                    totalSeasons.setText(tvSeriesModel.getNumberOfSeasons() + "S");
                    tagLine.setText(tvSeriesModel.getTagline());
                    if (tvSeriesModel.getBackdropPath() != null) {
                        Glide.with(getApplicationContext()).load("https://image.tmdb.org/t/p/w500" + tvSeriesModel.getBackdropPath()).into(backdrop);
                    } else {
                        Glide.with(getApplicationContext()).load("").into(backdrop);
                    }
                    recyclerGenre.setAdapter(new GenreAdapter(getApplicationContext(), tvSeriesModel.getGenres()));
                }
            }

            @Override
            public void onFailure(Call<TvSeriesModel> call, Throwable t) {

            }
        });

        tabLayout.addTab(tabLayout.newTab());
        tabLayout.addTab(tabLayout.newTab());
        tabLayout.addTab(tabLayout.newTab());


        TabLayoutAdapter tabAdapter = new TabLayoutAdapter(getSupportFragmentManager(), getApplicationContext(), tabLayout.getTabCount(), id, TabLayoutAdapter.SERIES);
        viewPager.setAdapter(tabAdapter);
        tabLayout.setupWithViewPager(viewPager);


    }
}

File: app/src/main/java/com/rkbapps/neetflix/activityes/TvSeriesPreviewActivity.kt
package com.rkbapps.neetflix.activityes

import android.os.Bundle
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.rkbapps.neetflix.R
import com.rkbapps.neetflix.adapter.GenreAdapter
import com.rkbapps.neetflix.adapter.tab.TabLayoutAdapter
import com.rkbapps.neetflix.databinding.ActivityTvSeriesPreviewBinding
import com.rkbapps.neetflix.db.Database
import com.rkbapps.neetflix.db.DatabaseReference.getDatabase
import com.rkbapps.neetflix.db.EntityModel
import com.rkbapps.neetflix.models.tvseries.TvSeriesModel
import com.rkbapps.neetflix.repository.series.TvSeriesPreviewRepository
import com.rkbapps.neetflix.services.RetrofitInstance
import com.rkbapps.neetflix.utils.ContentType
import com.rkbapps.neetflix.viewmodelfactories.series.TvSeriesPreviewViewModelFactory
import com.rkbapps.neetflix.viewmodels.series.TvSeriesPreviewViewModel
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch

class TvSeriesPreviewActivity : AppCompatActivity() {

    private lateinit var tvSeriesModel: TvSeriesModel
    private lateinit var binding: ActivityTvSeriesPreviewBinding
    private lateinit var viewModel: TvSeriesPreviewViewModel
    private lateinit var mDatabase: Database

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_tv_series_preview)
        mDatabase = getDatabase(this)
        val id = intent.getIntExtra("id", -1)
        val tittle = intent.getStringExtra("tittle")
        val repository = TvSeriesPreviewRepository(RetrofitInstance.tvSeriesApi!!)
        viewModel =
            ViewModelProvider(
                this,
                TvSeriesPreviewViewModelFactory(repository, id)
            )[TvSeriesPreviewViewModel::class.java]

        val viewPager = binding.pager
        val tabLayout = binding.tabLayout

        binding.recyclerGenre.layoutManager =
            LinearLayoutManager(this, RecyclerView.HORIZONTAL, false)

        binding.bookmarkSeries.isEnabled = false


        if (mDatabase.contentDao.isBookmarked(id)) {
            binding.bookmarkSeries.setImageResource(R.drawable.bookmark)
        } else {
            binding.bookmarkSeries.setImageResource(R.drawable.bookmark_border)
        }


        setSupportActionBar(binding.toolbarTvSeries)
        supportActionBar!!.title = tittle
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        loadTvSeriesData(id)

        tabLayout.addTab(tabLayout.newTab())
        tabLayout.addTab(tabLayout.newTab())
        tabLayout.addTab(tabLayout.newTab())
        val tabAdapter = TabLayoutAdapter(
            supportFragmentManager,
            applicationContext,
            tabLayout.tabCount,
            id,
            TabLayoutAdapter.SERIES
        )
        viewPager.adapter = tabAdapter
        tabLayout.setupWithViewPager(viewPager)

        binding.bookmarkSeries.setOnClickListener {
            if (mDatabase.contentDao.isBookmarked(id)) {
                MainScope().launch {
                    mDatabase.contentDao.removeBookmark(id)
                }
                Toast.makeText(
                    this@TvSeriesPreviewActivity,
                    "Removed from bookmark",
                    Toast.LENGTH_SHORT
                ).show()
                binding.bookmarkSeries.setImageResource(R.drawable.bookmark_border)
            } else {
                try {
                    addToBookmark(tvSeriesModel)
                } catch (e: Exception) {
                    Toast.makeText(
                        this@TvSeriesPreviewActivity,
                        "Something went wrong",
                        Toast.LENGTH_SHORT
                    ).show()

                }
            }
        }
    }

    private fun addToBookmark(tvSeriesModel: TvSeriesModel) {
        val entityModel = EntityModel(
            type = ContentType.SERIES,
            contentId = 0,
            adult = tvSeriesModel.adult,
            id = tvSeriesModel.id,
            releaseDate = tvSeriesModel.firstAirDate,
            posterPath = tvSeriesModel.posterPath,
            title = tvSeriesModel.name,
            voteAverage = tvSeriesModel.voteAverage
        )
        MainScope().launch {
            mDatabase.contentDao.addToBookmark(entityModel)
        }
        binding.bookmarkSeries.setImageResource(R.drawable.bookmark)
        Toast.makeText(
            this@TvSeriesPreviewActivity,
            "Added to bookmark",
            Toast.LENGTH_SHORT
        ).show()

    }


    private fun loadTvSeriesData(id: Int) {

        viewModel.seriesDetails.observe(this) {
            if (it != null) {
                tvSeriesModel = it
                binding.recyclerGenre.adapter =
                    GenreAdapter(this@TvSeriesPreviewActivity, tvSeriesModel.genres)

                binding.tvSeriesModel = tvSeriesModel

                binding.bookmarkSeries.isEnabled = true
            } else {
                Toast.makeText(
                    this@TvSeriesPreviewActivity,
                    "Something went wrong",
                    Toast.LENGTH_SHORT
                ).show()
            }
        }

//        RetrofitInstance.tvSeriesApi!!.getSeriesDetails(id, ApiData.API_KEY)
//            ?.enqueue(object : Callback<TvSeriesModel?> {
//                override fun onResponse(
//                    call: Call<TvSeriesModel?>,
//                    response: Response<TvSeriesModel?>,
//                ) {
//                    if (response.isSuccessful) {
//                        if (response.body() != null) {
//                            tvSeriesModel = response.body()!!
//                            binding.recyclerGenre.adapter =
//                                GenreAdapter(this@TvSeriesPreviewActivity, tvSeriesModel.genres)
//
//                            binding.tvSeriesModel = tvSeriesModel
//
//                            binding.bookmarkSeries.isEnabled = true
//                        }
//                    }
//                }
//
//                override fun onFailure(call: Call<TvSeriesModel?>, t: Throwable) {
//                    Toast.makeText(
//                        this@TvSeriesPreviewActivity,
//                        t.localizedMessage!!.toString(),
//                        Toast.LENGTH_SHORT
//                    ).show()
//                }
//            })
    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        onBackPressedDispatcher.onBackPressed()
        return true
    }
}